### PR TITLE
Conversions between CStr, OsStr, Path and boxes

### PR DESCRIPTION
**Pull request by clarfonthey**
This closes a bit of the inconsistencies between `CStr`, `OsStr`, `Path`, and `str`, allowing people to create boxed versions of DSTs other than `str` and `[T]`.

Full list of additions:
* `Default` for `Box<str>`, `Box<CStr>`, `Box<OsStr>`
* `CString::into_boxed_c_str` (feature gated)
* `OsString::into_boxed_os_str` (feature gated)
* `Path::into_boxed_path` (feature gated)
* `From<&CStr> for Box<CStr>`
* `From<&OsStr> for Box<OsStr>`
* `From<&Path> for Box<Path>`

This also includes adding the internal methods:
* `sys::*::os_str::Buf::into_box`
* `sys::*::os_str::Slice::{into_box, empty_box}`
* `sys_common::wtf8::Wtf8Buf::into_box`
* `sys_common::wtf8::Wtf8::{into_box, empty_box}`

### CODE REVIEW

File: src/libstd/ffi/c_str.rs
Comment (bjorn3) [inline src/libstd/ffi/c_str.rs]: I think this should be unstable.

**Review by bjorn3 — CHANGES_REQUESTED** [review]

File: src/libstd/ffi/c_str.rs
Comment (clarcharr) [inline src/libstd/ffi/c_str.rs]: Trait implementations are inherently stable, so, that's not an option.

**Comment by alexcrichton** [comment]
I [requested](https://github.com/rust-lang/rust/pull/39438#issuecomment-277783450) that this be split off #39438 (thanks @clarcharr!) because we don't have a lot of precedent for this in the standard library with other DSTs like `Path`, `OsStr`, etc. These seem reasonable to me, however, and I'd be fine adding all of them all at once as well.

**Comment by clarcharr** [comment]
I will add this for `OsStr` soon, although I need to find out a proper way to reconcile how the platform-specific types should be laid out. `Path` will come easily after that.

**Comment by clarcharr** [comment]
I've added versions for `OsStr` and `Path`, and it'd be nice if someone took a quick glance over the code to make sure everything looks right. Because AppVeyor doesn't automatically do PRs at the moment, there might be a bug in the Windows code that won't be caught until bors tries to merge it.

**Comment by clarcharr** [comment]
@aturon/@alexcrichton, I'll open up a tracking issue for the `into_boxed_*` methods once I get a nod that those methods are desired. This will be a bit of an odd change because the `From` impls will be immediately stable but the `into_*` methods will not be. If it makes more sense to open an RFC for this I can do that.

**Comment by alexcrichton** [comment]
Nah yeah we can deal with the tracking issue just before merging, and these impls shouldn't require an RFC. Thanks for the updates!

**Comment by clarcharr** [comment]
Currently waiting on #39438 which, even though it said it was merged before, was not actually merged.

**Comment by clarcharr** [comment]
I accidentally forgot to do a git pull before merging, so, I'll have to grab dec1e6423fd96fdf5409cbb53da6634913b148de from my other machine before we merge this.

**Comment by clarcharr** [comment]
Waiting on CI, but I've implemented everything necessary for this and made a full list in the OP of what was implemented.

**Comment by clarcharr** [comment]
Ping @aturon

**Comment by aturon** [comment]
@clarcharr Thanks! This looks good to me.

@rust-lang/libs: This PR involves some instantly-stable trait implementations. However, everything being implemented here is filling out our inter-conversion story to match other types in `std`, which should make things more consistent and predictable. I will likely r+ soon, so if you have objections or want time to consider further, please speak up!

**Comment by aturon** [comment]
@clarcharr One thing that makes me a bit uncomfortable about the implementation here is the pervasive use of `mem::transmute`. I wonder if we can at least factor out some common conversions into some free-standing functions that can be re-used? Each transmute is basically making an assertion about the relative representations, and it'd be good to consolidate those if we can.

**Comment by BurntSushi** [comment]
I think I'm fine with this.

In the implementation there are a lot of `transmute`s. Are they all necessary or is there anyway to get rid of some of them?

**Comment by BurntSushi** [comment]
@aturon Haha. Beat me to it. :P

**Comment by alexcrichton** [comment]
Looks good to me! I think we could annotate the input/output types in the transmutes, but otherwise I think they may unfortunately be required

**Comment by clarcharr** [comment]
The main issue with the transmutes is that there's no other way of converting newtype DSTs. I decided to do one type at a time which is why there are so many, because each transmute adds one more layer of newtypes.

It's most complicated in the instance of Path on Windows, which is [u8] -> Wtf8 -> Slice -> OsStr -> Path.

I *could* just coerce an empty u8 slice to a path, but that makes it harder to understand the conversion that's actually happening.

**Comment by aturon** [comment]
Alright, given that the `transmute`s appear necessary, let's go ahead and land this!

@bor: r+

**Comment by frewsxcv** [comment]
@bors r-

This fails to compile on Windows: https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1944/job/cu5i3oo7m5dq2hbg

**Comment by clarcharr** [comment]
Should be working now.

**Comment by jonhoo** [comment]
Out of curiosity, why is `into_boxed_c_str` behind a feature flag for now?

**Comment by clarcharr** [comment]
@jonhoo, by protocol a tracking issue is opened for every new feature. So you'll have to wait until this goes stable for that to happen.

**Comment by clarcharr** [comment]
I also realised that I never made a tracking issue for this, so, feel free to open a PR to add one in so that this can be nominated faster.

**Pull request merged**
